frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3956,6 +4011,145 @@ bool isTypeDefaultInitializable(Context* context, const Type* t) {
   return isTypeDefaultInitializableQuery(context, t);
 }
 
+void getCopyOrAssignableInfo(Context* context, const Type* t,
+                                    bool* fromConst, bool* fromRef,
+                                    bool checkCopyable);
+
+// checkCopyable is true for copyable, false for assignable.
+// Return: First is from const, second is from ref.
+static const std::pair<bool, bool>& getCopyOrAssignableInfoQuery(
+    Context* context, const Type* t, bool checkCopyable) {
+  QUERY_BEGIN(getCopyOrAssignableInfoQuery, context, t, checkCopyable);
+
+  CHPL_ASSERT((t->isClassType() || t->isCompositeType()) &&
+              "should only be getting called for class or composite types");
+
+  // Inspect type for either kind of copyability/assignability.
+  bool fromConst = false;
+  bool fromRef = false;
+  if (auto classTy = t->toClassType()) {
+    if (classTy->decorator().isManaged() &&
+        classTy->manager()->isAnyOwnedType()) {
+      // Owned class types are copyable from ref iff they are nilable.
+      fromRef = classTy->decorator().isNilable();
+    } else {
+      // Class types of other management are copyable from const.
+      fromConst = true;
+    }
+  } else if (auto ct = t->toCompositeType()) {
+    if (auto at = t->toArrayType()) {
+      if (auto eltType = at->eltType().type()) {
+        // Arrays are copyable/assignable if their elements are
+        getCopyOrAssignableInfo(context, eltType, &fromConst, &fromRef,
+                                checkCopyable);
+      }
+    } else if (auto tt = t->toTupleType()) {
+      // Tuples have the minimum copyable/assignable-ness of their elements
+      if (tt->isStarTuple()) {
+        getCopyOrAssignableInfo(context, tt->elementType(0).type(), &fromConst,
+                                &fromRef, checkCopyable);
+      } else {
+        bool allEltsFromConst = true;
+        bool allEltsFromRef = true;
+        for (int i = 0; i < tt->numElements(); i++) {
```

Comment:
Could you create and use an iterator for TupleType's element types, or list it as Future Work? It is not a big deal for this PR but IMO it'd make this section a fair amount nicer.

Reply:
Will do. I was thinking about that for `ResolvedFields` below as well so I'll probably just get both.

Reply:
This PR is starting to drag on a bit for me, so maybe I will just leave it off to future work after all